src/org/jgroups/raft/util/LogCache.java

```diff
@@ -102,38 +101,58 @@ public void append(int index, boolean overwrite, LogEntry... entries) {
         log.append(index, overwrite, entries);
         last_appended=log.lastAppended();
         current_term=log.currentTerm();
-        for(int i=0; i < entries.length; i++)
-            cache.put(index++, entries[i]);
+        for (int i = 0; i < entries.length; i++) {
+            final int logIndex = index + i;
+            if (logIndex >= cache.getHeadSequence()) {
+                if (cache.availableCapacityWithoutResizing() == 0) {
+                    // try trim here to see if we can save enlarging to happen
+                    trim();
+                }
+                cache.set(logIndex, entries[i]);
+            }
+        }
         trim();
     }
 
     public LogEntry get(int index) {
         if(index > last_appended) // called by every append() to check if the entry is already present
             return null;
-
-        LogEntry e=cache.get(index);
-        if(e != null) {
-            num_hits++;
-            return e;
+        if (index < cache.getHeadSequence()) {
```

Comment:
@belaban 
This is the part that requires attention: on reload I expect "someone" to move the head sequence of the ring buffer to an higher value (== commit index + 1). 
I think we should discuss the easier way to handle that; I can even add a method (moveHead) to the ring buffer; let's decide together what's best